In [12]:
from brainlit.map_neurons.map_neurons import (
    DiffeomorphismAction,
    transform_geometricgraph,
    compute_derivs,
    CloudReg_Transform,
)
from brainlit.algorithms.trace_analysis.fit_spline import (
    GeometricGraph,
    compute_parameterization,
    CubicHermiteChain,
)
from brainlit.utils.Neuron_trace import NeuronTrace

from pathlib import Path
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.cm import ScalarMappable
from mpl_toolkits.mplot3d import Axes3D
import pandas as pd
from scipy.interpolate import splev, splprep, CubicHermiteSpline
from scipy.spatial.distance import cosine
from scipy.stats import wilcoxon, pearsonr, linregress, norm
from tqdm import tqdm
from cloudvolume import CloudVolume
import seaborn as sns
from cloudvolume.exceptions import SkeletonDecodeError
import pickle
from copy import deepcopy
import networkx as nx
import os

In [13]:
root_dir = Path(os.path.abspath(""))
data_dir = os.path.join(root_dir, "data", "mapping-files")

In [14]:
velocity_path = os.path.join(data_dir, "downloop_1_v.mat")
affine_path = os.path.join(data_dir, "downloop_1_A.mat")

ct = CloudReg_Transform(velocity_path, affine_path)

integrating velocity field: 100%|██████████| 10/10 [00:26<00:00,  2.60s/it]


In [15]:
axons_path = "precomputed://file://" + os.path.join(data_dir, "axons")
vol = CloudVolume(axons_path)
shp = np.array(vol.shape)
res = np.array(vol.resolution) / 1000
origin = (shp[:3] - 1) * res / 2

skel = vol.skeleton.get(4)
coords = skel.vertices / 1000 - origin
coords = ct.apply_affine(coords)

## original vertices
The following are the vertex coordinates stored in the skeleton file

In [16]:
coords_skel = skel.vertices
print(coords_skel)

[[6680812.5 4213735.5 5570578. ]
 [6679242.  4213755.  5572043. ]
 [6660847.5 4216505.  5564594. ]
 ...
 [7051972.5 3908219.8 4851556.5]
 [7074937.5 4341430.5 5281373. ]
 [7068199.  4344958.  5270404.5]]


## neuroglancer
these are the voxel coordinates (of the highest res image) that you would use in neuroglancer

In [17]:
coords_ng = np.divide(coords_skel, 1000 * np.array([19.2, 19.2, 64]))
print(coords_ng)

[[347.95898438 219.46539062  87.04028125]
 [347.8771875  219.46640625  87.06317187]
 [346.91914063 219.60963542  86.94678125]
 ...
 [367.29023438 203.55311198  75.80557031]
 [368.48632812 226.11617187  82.52145312]
 [368.13536458 226.29989583  82.35007031]]


## centered
now, we changed units to microns and centered the image

In [18]:
coords_centered = coords_skel / 1000 - origin
print(coords_centered)

[[1468.0125      229.73535156  482.578125  ]
 [1466.4421875   229.75488281  484.04296875]
 [1448.04765625  232.50488281  476.59423828]
 ...
 [1839.17265625  -75.78027344 -236.44335938]
 [1862.1375      357.43066406  193.37304688]
 [1855.39921875  360.95800781  182.40429688]]


## affine
applied affine transform to coordinates

In [19]:
coords_affine = ct.apply_affine(coords_centered)
print(coords_affine)

[[ 524.30916956  737.44099749 1371.16397248]
 [ 525.84386745  737.47461986 1369.48174897]
 [ 518.37428854  739.81165879 1349.2803704 ]
 ...
 [-210.42621759  335.60447696 1773.2164223 ]
 [ 214.64419038  898.37724199 1790.00024407]
 [ 203.32006118  901.97601536 1782.37493956]]


## complete transform

In [20]:
coords_transform = ct.evaluate(coords_affine)
print(coords_transform)

[[ 528.58474024  736.84231508 1368.12984341]
 [ 530.12180605  736.87501546 1366.44977923]
 [ 522.66172355  739.19192612 1346.27765454]
 ...
 [-206.73472768  336.05556287 1770.35923739]
 [ 218.37173767  897.91184161 1786.17772276]
 [ 207.04544281  901.49831338 1778.55821141]]


## ng atlas voxel space

In [21]:
coords_ng_atlas = np.divide(coords_transform, 10) + np.array([1320, 800, 1140]) / 2
coords_ng_atlas

array([[712.85847402, 473.68423151, 706.81298434],
       [713.01218061, 473.68750155, 706.64497792],
       [712.26617235, 473.91919261, 704.62776545],
       ...,
       [639.32652723, 433.60555629, 747.03592374],
       [681.83717377, 489.79118416, 748.61777228],
       [680.70454428, 490.14983134, 747.85582114]])

## cloudreg transform point process summary:
- voxel coordinates
- convert to microns
- center (target) image

## result: 
[528.6, 736.8, 1368.1] vs cloudreg [528.8, 737.2, 1368.6]